test/cri-containerd/stats_test.go

```diff
@@ -0,0 +1,152 @@
+	}
+	if stat.Memory.WorkingSetBytes.Value == 0 {
+		t.Fatalf("expected memory usage != 0 but got %d", stat.Memory.WorkingSetBytes.Value)
+	}
+}
+
+func Test_UVMStats_List_ContainerID_LCOW(t *testing.T) {
+	pullRequiredLcowImages(t, []string{imageLcowK8sPause})
+
+	request := &runtime.RunPodSandboxRequest{
+		Config: &runtime.PodSandboxConfig{
+			Metadata: &runtime.PodSandboxMetadata{
+				Name:      t.Name(),
+				Namespace: testNamespace,
+			},
+		},
+		RuntimeHandler: lcowRuntimeHandler,
+	}
+
+	client := newTestRuntimeClient(t)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	podID := runPodSandbox(t, client, ctx, request)
+	defer stopAndRemovePodSandbox(t, client, ctx, podID)
+
+	statsRequest := &runtime.ListContainerStatsRequest{
+		Filter: &runtime.ContainerStatsFilter{
+			Id: podID,
+		},
+	}
+
+	stats, err := client.ListContainerStats(ctx, statsRequest)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(stats.Stats) != 1 {
+		t.Fatalf("expected 1 stats result but got %d", len(stats.Stats))
+	}
+	stat := stats.Stats[0]
+	if stat.Cpu.Timestamp == 0 {
+		t.Fatalf("expected cpu stat timestamp != 0 but got %d", stat.Cpu.Timestamp)
+	}
+	if stat.Cpu.UsageCoreNanoSeconds.Value == 0 {
+		t.Fatalf("expected cpu usage != 0 but got %d", stat.Cpu.UsageCoreNanoSeconds.Value)
+	}
+	if stat.Memory.Timestamp == 0 {
+		t.Fatalf("expected memory stat timestamp != 0 but got %d", stat.Memory.Timestamp)
+	}
+	if stat.Memory.WorkingSetBytes.Value == 0 {
+		t.Fatalf("expected memory usage != 0 but got %d", stat.Memory.WorkingSetBytes.Value)
+	}
+}
+
+func Test_UVMStats_List_PodID_LCOW(t *testing.T) {
+	pullRequiredLcowImages(t, []string{imageLcowK8sPause})
+
+	request := &runtime.RunPodSandboxRequest{
+		Config: &runtime.PodSandboxConfig{
+			Metadata: &runtime.PodSandboxMetadata{
+				Name:      t.Name(),
+				Namespace: testNamespace,
+			},
+		},
+		RuntimeHandler: lcowRuntimeHandler,
+	}
+
+	client := newTestRuntimeClient(t)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	podID := runPodSandbox(t, client, ctx, request)
+	defer stopAndRemovePodSandbox(t, client, ctx, podID)
+
+	statsRequest := &runtime.ListContainerStatsRequest{
+		Filter: &runtime.ContainerStatsFilter{
+			PodSandboxId: podID,
+		},
+	}
+
+	stats, err := client.ListContainerStats(ctx, statsRequest)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(stats.Stats) != 1 {
+		t.Fatalf("expected 1 stats result but got %d", len(stats.Stats))
+	}
+	stat := stats.Stats[0]
+	if stat.Cpu.Timestamp == 0 {
+		t.Fatalf("expected cpu stat timestamp != 0 but got %d", stat.Cpu.Timestamp)
+	}
+	if stat.Cpu.UsageCoreNanoSeconds.Value == 0 {
+		t.Fatalf("expected cpu usage != 0 but got %d", stat.Cpu.UsageCoreNanoSeconds.Value)
+	}
+	if stat.Memory.Timestamp == 0 {
+		t.Fatalf("expected memory stat timestamp != 0 but got %d", stat.Memory.Timestamp)
+	}
+	if stat.Memory.WorkingSetBytes.Value == 0 {
```

Comment:
Want to add a test with physical memory as well then you can verify it exactly?

Reply:
Upon investigation, it appears this approach to calculating working set only works for VA backed...

I created a pod with the following annotations:

```
    "annotations": {
        "io.microsoft.virtualmachine.computetopology.memory.allowovercommit": "false",
        "io.microsoft.virtualmachine.computetopology.memory.sizeinmb": "2048"
    },
```

And querying stats shows a working set of ~28MB, when I would expect it to be 2GB. I'm guessing to support stats for a physical backed UVM we will need to get the value some other way (perhaps we can actually use the HCS query for this).

Reply:
And were sure this is the correct number for VA then?

Reply:
You're right that I should do some more validation with a guest workload that uses memory to verify this. When I talked to Duen-wen this approach was believed to work, but I'll try to validate that.

Reply:
Okay, I used shimdiag to dd 500MB from /dev/zero to /tmp/foo in the UVM (/tmp is a tmpfs), and I see the working set size grow by 500MB. So I think it's safe to say the working set measurement works for VA backed.

Reply:
I also tried running `./shimdiag exec $s sh -c 'yes > /dev/null'` and then checking CPU stats via crictl, and see the CPU usage reported as ~200%. I assume crictl is not adjusting for the two cores on the UVM in its calculation, so I think this looks good.

Reply:
Sweet thanks!